### ISSUE TITLE
S2259 FP: SingleOrDefault() and FirstOrDefault() used within EF LINQ queries

### ISSUE THREAD

**Issue by sychare** [SonarSource/sonar-dotnet#6176]
### Description

Latest update causes false-positives on SingleOrDefault() and FirstOrDefault() used within EF core LINQ queries.

Rule ID S2259.

I suspect this is related to #3290.

### Repro steps

Within an Entity Framework core Select() statement it is valid to use SingleOrDefault()/FirstOrDefault() as the EF provider simply chains a null value onwards within the navigation if the specific item is null, therefore not throwing an NRE

```csharp
_efContext.MyTableName
    .Select(
        s =>
            new SomeResultObject
            {
                Id = s.Id,
                SomethingNullable = s.SomeOtherTableThatMightHaveOneOrNoRows.SingleOrDefault().ColumnInOtherTable
            }
    );
```

If no row exists within `SomeOtherTableThatMightHaveOneOrNoRows` then `SomethingNullable` will simply be null, not throw.

### Expected behavior

No violation of S2259.

### Actual behavior

False positive of S2259.

### Known workarounds

Manually mark as false positive.

### Related information

SonarCloud.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @sychare,
I'm not EF expert so I might need a bit of help here. If I understand it right:
* `_efContext.SomeOtherTableThatMightHaveOneOrNoRows.SingleOrDefault().ColumnInOtherTable` would throw an exception
* `s.SomeOtherTableThatMightHaveOneOrNoRows.SingleOrDefault().ColumnInOtherTable` doesn't throw, because it's passed to `.Select` as an `Linq.Expressions.Expression`

Is that correct?

Can you please show me how `MyTableName.SomeOtherTableThatMightHaveOneOrNoRows` property is declared in your code?

**Comment by sychare** [comment]
Hi @pavel-mikula-sonarsource, I believe your understanding is correct, yes.

`_efContext.SomeOtherTableThatMightHaveOneOrNoRows.SingleOrDefault().ColumnInOtherTable` _could_ throw an exception, as at the point of calling `SingleOrDefault()` the query itself would be translated to SQL and executed, therefore if there were not a single row in `SomeOtherTableThatMightHaveOneOrNoRows` then `.ColumnInOtherTable` would be dereferencing a null object.

The above is true only because there is no `Select()` projection here. It is within the various projection (e.g. `Select()`) and constraint clasues (e.g `Where()`) that the syntax won't throw. Not sure of an exhaustive list.

`MyTableName.SomeOtherTableThatMightHaveOneOrNoRows` is declared (as scaffolded using `dotnet ef dbcontext scaffold`) on `MyTableName` as:
```csharp
public virtual ICollection<SomeOtherTableThatMightHaveOneOrNoRows> SomeOtherTableThatMightHaveOneOrNoRows { get; set; }
```

**Comment by pavel-mikula-sonarsource** [comment]
Thanks @sychare, that should be enough information to fix the problem.

ToDo: Do not analyze arguments of `Linq.Expressions.Expression` with Symbolic Execution
